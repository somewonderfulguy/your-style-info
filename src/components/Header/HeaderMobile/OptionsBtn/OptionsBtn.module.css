:root {
  --mobile-menu-options-size: 20px;
  --mobile-menu-options-dot-diameter: 4px;
}

.btn {
  position: relative;
  width: var(--mobile-menu-options-size);
  height: var(--mobile-menu-options-size);

  display: flex;
  justify-content: center;
  
  outline: none;
}

.darkBtn { composes: btn; }

.dot {
  width: var(--mobile-menu-options-dot-diameter);
  height: var(--mobile-menu-options-dot-diameter);

  background-color: #000;
  border-radius: 50%;

  position: absolute;
  transition: background-color .2s;
}

.darkBtn .dot {
  background-color: lightgray;
}

.dot:nth-child(1) {top: 0; transform: rotate(-45deg);}
.dot:nth-child(2) {top: calc(50% - var(--mobile-menu-options-dot-diameter) / 2);}
.dot:nth-child(3) {bottom: 0; transform: rotate(45deg);}