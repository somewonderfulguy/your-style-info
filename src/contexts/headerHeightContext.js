import React, {createContext, useContext, useMemo, useState} from 'react'

const HeaderHeightContext = createContext()

const useHeaderHeight = () => {
  const context = useContext(HeaderHeightContext)
  if(!context) {
    throw new Error('useHeaderHeight must be used within a HeaderHeightProvider')
  }
  return context
}

const HeaderHeightProvider = props => {
  const [headerHeight, setHeaderHeight] = useState(0)
  const value = useMemo(() => ({headerHeight, setHeaderHeight}), [headerHeight])
  return <HeaderHeightContext.Provider value={value} {...props} />
}

export {HeaderHeightProvider, useHeaderHeight}