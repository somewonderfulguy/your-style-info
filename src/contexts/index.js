export * from './headerHeightContext'
export * from './loadingContext'
export * from './localisationContext'
export * from './screenDimensionsContext'
export * from './themeContext'