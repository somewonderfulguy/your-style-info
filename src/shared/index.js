export * from './componentRenderer'
export * from './getComponent'
export * from './tests'