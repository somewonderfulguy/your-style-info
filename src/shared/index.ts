export * from './hooks'
export * from './tests'
export * from './utils'

export * from './types'