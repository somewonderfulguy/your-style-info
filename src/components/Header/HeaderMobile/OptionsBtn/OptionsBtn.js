import React, {useRef} from 'react'
import {bool, func} from 'prop-types'
import {useSpring, animated} from 'react-spring'

import {useTheme} from 'helpers/contexts'
import styles from './OptionsBtn.module.css'

const propTypes = {
  isOpen: bool,
  onClick: func
}

const defaultProps = {
  isOpen: false,
  onClick: () => {}
}

const OptionsBtn = ({isOpen, onClick}) => {
  const optionsBtnDOM = useRef(null)
  const {isDarkTheme} = useTheme()
  const optionsHeight = optionsBtnDOM.current ? optionsBtnDOM.current.clientHeight : 0

  // TODO delete this, awful idea
  const dotSize = +(
    getComputedStyle(document.documentElement).getPropertyValue('--mobile-menu-options-dot-diameter')
      .replace('px', '')
      || 4
  )
  const shrinkedDotSize = dotSize / 2

  const middlePosition = (optionsHeight / 2) - shrinkedDotSize / 2

  const {timeline} = useSpring({
    config: {duration: 300},
    timeline: isOpen ? 1 : 0
  })

  const range = [0, 0.4, 0.6, 1]

  const topDot = {
    top: timeline.interpolate(range, [0, middlePosition, middlePosition, middlePosition]),
    width: timeline.interpolate(range, [4, 2, 2, 24]),
    height: timeline.interpolate(range, [4, 2, 2, 2]),
    borderRadius: timeline.interpolate(range, ['50%', '0', '0', '0'])
  }

  const bottomDot = {
    bottom: topDot.top,
    width: topDot.width,
    height: topDot.height,
    borderRadius: timeline.interpolate(range, ['50%', '0', '0', '0'])
  }

  const middleDot = {
    opacity: timeline.interpolate({range, output: [1, 0, 0, 0]})
  }

  return (
    <button ref={optionsBtnDOM} onClick={onClick} className={styles[isDarkTheme ? 'darkBtn' : 'btn']} type="button">
      <animated.div className={styles.dot} style={topDot} />
      <animated.div className={styles.dot} style={middleDot} />
      <animated.div className={styles.dot} style={bottomDot} />
    </button>
  )
}

OptionsBtn.propTypes = propTypes
OptionsBtn.defaultProps = defaultProps

export default OptionsBtn