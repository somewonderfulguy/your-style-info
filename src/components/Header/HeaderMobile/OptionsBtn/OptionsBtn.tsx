import React, { forwardRef, useImperativeHandle, useRef } from 'react'
import { useSpring, animated } from 'react-spring'

import { useThemeState } from 'contexts'
import styles from './OptionsBtn.module.css'

type propType = {
  isOpen?: boolean
  onClick?: () => void
}

const OptionsBtn = forwardRef(
  ({ isOpen = false, onClick = () => {} }: propType, ref) => {
    const optionsBtnDOM = useRef<HTMLButtonElement>(null)
    useImperativeHandle(ref, () => ({ optionsBtnDOM }))

    const isDarkTheme = useThemeState()

    const optionsHeight = optionsBtnDOM.current
      ? optionsBtnDOM.current.clientHeight
      : 0

    const dotSize = 4
    const shrinkedDotSize = dotSize / 2
    const middlePosition = optionsHeight / 2 - shrinkedDotSize / 2

    const { timeline: t } = useSpring({
      config: { duration: 300 },
      timeline: isOpen ? 1 : 0
    })
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const timeline = t as any

    const range = [0, 0.4, 0.6, 1]

    const topDot = {
      top: timeline.interpolate(range, [
        0,
        middlePosition,
        middlePosition,
        middlePosition
      ]),
      width: timeline.interpolate(range, [
        dotSize,
        shrinkedDotSize,
        shrinkedDotSize,
        24
      ]),
      height: timeline.interpolate(range, [
        dotSize,
        shrinkedDotSize,
        shrinkedDotSize,
        2
      ]),
      borderRadius: timeline.interpolate(range, ['50%', '0', '0', '0'])
    }

    const bottomDot = {
      bottom: topDot.top,
      width: topDot.width,
      height: topDot.height,
      borderRadius: timeline.interpolate(range, ['50%', '0', '0', '0'])
    }

    const middleDot = {
      width: dotSize,
      height: dotSize,
      opacity: timeline.interpolate({ range, output: [1, 0, 0, 0] })
    }

    return (
      <button
        ref={optionsBtnDOM}
        onClick={onClick}
        className={styles[isDarkTheme ? 'darkBtn' : 'btn']}
        type="button"
        // TODO: do it dynamic
        title="Options"
      >
        <animated.div className={styles.dot} style={topDot} />
        <animated.div className={styles.dot} style={middleDot} />
        <animated.div className={styles.dot} style={bottomDot} />
      </button>
    )
  }
)
OptionsBtn.displayName = 'OptionsBtn'

export default OptionsBtn
