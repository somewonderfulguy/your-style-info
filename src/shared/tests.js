import React from 'react'
import {render as rtlRender} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {Router} from 'react-router-dom'
import {createMemoryHistory} from 'history'
import {QueryClient, QueryClientProvider} from 'react-query'

import {HeaderHeightProvider, LocalizationProvider, ScreenDimensionsProvider, ThemeProvider} from 'contexts'

const queryClient = new QueryClient({
  defaultOptions: {
    initialStale: true
  }
})

const render = (ui, {route = '/', ...options} = {}) => {
  const history = createMemoryHistory({initialEntries: [route]})
  const Wrapper = props => (
    <QueryClientProvider client={queryClient}>
      <Router history={history}>
        <HeaderHeightProvider>
          <LocalizationProvider>
            <ThemeProvider>
              <ScreenDimensionsProvider {...props} />
            </ThemeProvider>
          </LocalizationProvider>
        </HeaderHeightProvider>
      </Router>
    </QueryClientProvider>
  )
  return {
    ...rtlRender(ui, {wrapper: Wrapper, ...options}),
    history,
  }
}

const renderWithoutProviders = rtlRender

export * from '@testing-library/react'
export {render, renderWithoutProviders, userEvent}