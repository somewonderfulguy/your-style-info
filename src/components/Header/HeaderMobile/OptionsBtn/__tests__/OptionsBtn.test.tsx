import React, { FunctionComponent } from 'react'
import { render, screen, waitFor } from '@testing-library/react'

import { ThemeProvider } from 'contexts'
import OptionsBtn from '..'
import { renderWholeApp, userEvent } from 'shared/tests'

const setup = () =>
  render(<OptionsBtn />, { wrapper: ThemeProvider as FunctionComponent })

test('snapshot diffrenece between default and clicked states', async () => {
  const { container, rerender } = setup()
  const dots = container.querySelectorAll('.dot')

  const defaultStyles = [
    { top: '0px', width: '4px', height: '4px', borderRadius: '50%' },
    { width: '4px', height: '4px', opacity: '1' },
    { bottom: '0px', width: '4px', height: '4px', borderRadius: '50%' }
  ]
  const clickedStyles = [
    { top: '-1px', width: '24px', height: '2px' },
    { width: '4px', height: '4px', opacity: '0' },
    { bottom: '-1px', width: '24px', height: '2px' }
  ]

  // default state asserts
  dots.forEach((dot, idx) => expect(dot).toHaveStyle(defaultStyles[idx]))

  rerender(<OptionsBtn isOpen />)

  // clicked state asserts
  await waitFor(() =>
    dots.forEach((dot, idx) => expect(dot).toHaveStyle(clickedStyles[idx]))
  )
})

test('snapshot difference in dark/light theme', () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  window.innerWidth = 320

  renderWholeApp()

  const optionsBtn = screen.getByTitle(/options/i)
  const lightThemeClass = optionsBtn.className

  userEvent.click(screen.getAllByLabelText(/switch theme/i)[0])

  const darkThemeClass = optionsBtn.className

  expect(lightThemeClass).toMatchDiffSnapshot(darkThemeClass, {
    contextLines: 0,
    aAnnotation: 'light theme',
    bAnnotation: 'dark theme'
  })
})
