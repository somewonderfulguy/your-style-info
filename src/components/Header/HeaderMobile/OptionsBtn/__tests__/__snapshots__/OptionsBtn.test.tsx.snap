// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot difference in dark/light theme 1`] = `
Snapshot Diff:
- light theme
+ dark theme

- btn
+ darkBtn
`;
